python/ray/data/dataset.py

```diff
@@ -369,15 +369,32 @@ def repartition(self, num_blocks: int, *,
         reduce_bar = ProgressBar("Repartition", position=0, total=len(splits))
         reduce_out = [
             reduce_task.remote(*s.get_internal_block_refs()) for s in splits
+            if s.num_blocks() > 0
```

Comment:
We shouldn't filter out empty blocks here; that can happen in general if read tasks / user map tasks generate empty output.

Instead, change groupby to be robust to empty blocks.

Reply:
@ericl To be clear, groupby is already robust to empty blocks, this is a driveby fix of `.repartition()` and is orthogonal to groupby. This wasn't previously covered by tests.

Without this fix in `.repartition()`, if this is a simple dataset (with simple blocks) and `num_blocks > ds.count()`, the created empty blocks will be Arrow blocks, and that mix of simple and Arrow blocks will break downstream operations. This is because the [reduce task gets an empty split with no blocks](https://github.com/ray-project/ray/blob/6bb0659010ade8031622ef38bc9fcfa048a71dcb/python/ray/data/dataset.py#L370-L372), the [reduce function builds the block with the delegating builder](https://github.com/ray-project/ray/blob/6bb0659010ade8031622ef38bc9fcfa048a71dcb/python/ray/data/impl/shuffle.py#L125-L128), which [defaults to an Arrow block if no blocks have been added](https://github.com/ray-project/ray/blob/6bb0659010ade8031622ef38bc9fcfa048a71dcb/python/ray/data/impl/arrow_block.py#L98-L99).

Ideally we could fix this in `_shuffle_reduce()`, but it's getting an empty list of blocks, so it has no idea what type the blocks are supposed to be.

Reply:
@ericl Resolving this thread, let me know if you have any other questions about this driveby fix of `.repartition()`.